app/controllers/submissions_controller.rb

```diff
@@ -44,47 +48,47 @@ def show
           user_id: @submission.user_id,
           course: @course.name,
           exercise_name: @submission.exercise.name,
-          status: @submission.status,
+          status: @submission.status(current_user),
           points: @submission.points_list,
+          validations: @submission.validations,
+          valgrind: @submission.valgrind,
+          solution_url: @exercise.solution.visible_to?(current_user) ? view_context.exercise_solution_url(@exercise) : nil,
+          submitted_at: @submission.created_at,
           processing_time: @submission.processing_time,
-          message_for_paste: @submission.message_for_paste,
+          reviewed: @submission.reviewed?,
+          requests_review:  @submission.requests_review?,
+          paste_url: @submission.paste_available ? paste_url(@submission.paste_key) : nil,
+          message_for_paste: @submission.paste_available ? @submission.message_for_paste : nil,
           missing_review_points: @exercise.missing_review_points_for(@submission.user)
         }
+
         output = output.merge(
-          case @submission.status
-          when :processing then {
-            submissions_before_this: @submission.unprocessed_submissions_before_this,
-            total_unprocessed: Submission.unprocessed_count
-          }
-          when :error then { error: @submission.pretest_error }
-          when :fail then {
-            test_cases: @submission.test_case_records
-          }
-          when :ok then {
-            test_cases: @submission.test_case_records,
-            feedback_questions: @course.feedback_questions.order(:position).map(&:record_for_api),
-            feedback_answer_url: submission_feedback_answers_url(@submission, format: :json),
-            processing_time: @submission.processing_time
-          }
+          case @submission.status(current_user)
+            when :processing then {
+              submissions_before_this: @submission.unprocessed_submissions_before_this,
+              total_unprocessed: Submission.unprocessed_count
+            }
+            when :ok then {
+              test_cases: @submission.test_case_records,
+              feedback_questions: @course.feedback_questions.order(:position).map(&:record_for_api),
+              feedback_answer_url: submission_feedback_answers_url(@submission, format: :json),
+            }
+            when :fail then {
+              test_cases: @submission.test_case_records
+            }
+            when :hidden then {
```

Comment:
this hidden feels like it the hidden field for course. I assume it really refers to the hide submission results. consider renaming it; now it could easily confuse endusers and developers